### PR TITLE
Ignore opcodes when attempting to fold array class tests

### PR DESCRIPTION
**Pull request by jdmpapin**
The previous logic here incorrectly assumed that an `aloadi` must load `<vft-symbol>`, but it could load from elsewhere, e.g. a field of a reflective object. When loading `<classFromJavaLangClass>`, the base object could be constrained with `isClassObject() == TR_yes`, which would result in the right type even though it doesn't seem like the previous logic was intended to take this case into account. But it could also be constrained with `isClassObject() != TR_yes`, and with a type bound. In that case, the type bound describes the type of the object we're loading from, which is actually as expected by the previous logic, but because the `aloadi` loads `<classFromJavaLangClass>` rather than `<vft-symbol>`, that type is not in general related to the result of the load, and VP could transform the tree incorrectly. More specifically, the test could be folded to 0 (indicating a non-array type) even though the reflectively identified type could turn out to be an array type at runtime.

The opcode of the node that produces the class pointer is irrelevant. Just get the constraint for that node, which always has a consistent interpretation, instead of trying to guess what `getClassType()` means based on multiple possible choices of `baseExpression`.

This change also generalizes the transformation to cases where the opcode is neither `aloadi` nor `loadaddr`. For example, the class pointer could come from a direct aload of a temporary, if we have a bound for the classes that the temporary could possibly point to. As such, there are no "unsuitable" cases to be counted, only hits and misses.

The remaining static debug counter increments were almost identical, and they have been consolidated.

The "Fold isArray test" trace message is removed because folding will already be plainly visible in logs due to the `performTransformation()`.

### CODE REVIEW

**Comment by klangman** [comment]
@0xdaryl Since Vijay is away, maybe you are able to review this one?

File: compiler/optimizer/VPHandlers.cpp [hunk 1]
```diff
@@ -7690,42 +7690,34 @@ TR::Node *constrainIand(OMR::ValuePropagation *vp, TR::Node *node)
...
+                   && classConstraint->isJ9ClassObject() == TR_yes
+                   && classConstraint->getClassType() != NULL
+                   && classConstraint->getClassType()->isArray() != TR_maybe)
+                  {
+                  hitOrMiss = "hit";
 
-                     TR::DebugCounter::incStaticDebugCounter(vp->comp(), TR::DebugCounter::debugCounterName(vp->comp(), "isArrayTest/hit/(%s)/%s", vp->comp()->signature(), vp->comp()->getHotnessName(vp->comp()->getMethodHotness())));
-                     }
+                  if (classConstraint->getClassType()->isArray() == TR_yes)
+                     constraint = TR::VPIntConst::create(vp, rhs->asIntConst()->getLowInt());
```
Comment (vijaysun-omr) [inline compiler/optimizer/VPHandlers.cpp:7708]: Can we just use the constant value 1 here ?

**Comment by vijaysun-omr** [comment]
This looks like it could be an improvement (for performance) but I suspect the reason it was done was as a functional bug fix. While I can see the reasoning from why it is an improvement from the comments and the code, it is not obvious to me how/if this is also needed for functional correctness. Could you please clarify ?

**Comment by jdmpapin** [comment]
I've updated the commit message (and initial PR comment) to expound on the functional problem fixed by this change

**Comment by vijaysun-omr** [comment]
Jenkins build all

**Review by vijaysun-omr — APPROVED** [review]

**Pull request merged**